### PR TITLE
Add gatsby-link snapshot testing

### LINKED ISSUE #7089: Add snapshot tests
## Summary

I'd like to add snapshot testing for some of the packages, like `gatsby-link`. Would also like to see more DOM tests on these kind of components with eg [`react-testing-library`](https://github.com/kentcdodds/react-testing-library) 

### Motivation

Better testing will lead to fewer bugs and will make it easier for new contributors to not break things

I'd be happy to create a PR adding this to `gatsby-link` but I wanted to hear what you think before.

### PR DESCRIPTION
**Pull request by alexandernanberg**
Fixes #7089

Right now this is a very simple setup of snapshots but can easily the extended in the future. With `react-testing-library` there is also the possibility of adding more DOM tests which is very nice.

Similar setup should be fairly simple to add to other packages as well

Noticed that `react-router-dom` is still a dependency but saw that there already is a PR for that so I won't touch that now.

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Deploy preview for *gatsbygram* failed.

Built with commit c843ac30b35e8baaf3a9c4491a0c3df73df0ea07

https://app.netlify.com/sites/gatsbygram/deploys/5b694e9add28ef0e8b98aec6

**Comment by KyleAMathews** [comment]
Deploy preview for *using-postcss-sass* failed.

Built with commit c843ac30b35e8baaf3a9c4491a0c3df73df0ea07

https://app.netlify.com/sites/using-postcss-sass/deploys/5b694e9bdd28ef0e8b98aecf

File: packages/gatsby-link/src/__tests__/index.js [hunk 1]
```diff
@@ -1,10 +1,13 @@
...
 import {
   createMemorySource,
   createHistory,
   LocationProvider,
 } from "@reach/router"
+import Link from "../"
+
+afterEach(cleanup)
 
 const getInstance = (props, pathPrefix = ``) => {
```
Comment (alexandernanberg) [inline packages/gatsby-link/src/__tests__/index.js]: Btw is this still relevant to test?

Comment (KyleAMathews) [reply packages/gatsby-link/src/__tests__/index.js]: Yeah — it lets tests change the `pathPrefix`. So tests should use it generally. It does need the context removed I'm noticing.

Comment (KyleAMathews) [reply packages/gatsby-link/src/__tests__/index.js]: Though it looks like it's easier to merge it into what you're doing

**Comment by KyleAMathews** [comment]
Deploy preview for *image-processing* failed.

Built with commit d5d7b31e0924c766bec7bf13b90760bffef4225e

https://app.netlify.com/sites/image-processing/deploys/5b69537f67610c5153d15965

**Comment by KyleAMathews** [comment]
Deploy preview for *using-remark* failed.

Built with commit d5d7b31e0924c766bec7bf13b90760bffef4225e

https://app.netlify.com/sites/using-remark/deploys/5b69538067610c5153d15971

**Comment by KyleAMathews** [comment]
Deploy preview for *using-contentful* failed.

Built with commit d5d7b31e0924c766bec7bf13b90760bffef4225e

https://app.netlify.com/sites/using-contentful/deploys/5b69538067610c5153d1596e

**Comment by KyleAMathews** [comment]
Deploy preview for *using-styled-components* failed.

Built with commit d5d7b31e0924c766bec7bf13b90760bffef4225e

https://app.netlify.com/sites/using-styled-components/deploys/5b69538067610c5153d1596b

**Comment by KyleAMathews** [comment]
Deploy preview for *using-glamor* failed.

Built with commit d5d7b31e0924c766bec7bf13b90760bffef4225e

https://app.netlify.com/sites/using-glamor/deploys/5b69537f67610c5153d15968

**Comment by KyleAMathews** [comment]
Deploy preview for *using-jss* failed.

Built with commit d5d7b31e0924c766bec7bf13b90760bffef4225e

https://app.netlify.com/sites/using-jss/deploys/5b69538367610c5153d15986

File: packages/gatsby-link/src/__tests__/index.js (hunk 1 above)
Comment (alexandernanberg) [reply packages/gatsby-link/src/__tests__/index.js]: Okay, so instead of using `import Link from '../'` we should use the `getInstance` if I understand it correctly?

Comment (KyleAMathews) [reply packages/gatsby-link/src/__tests__/index.js]: Importing is fine. It's just we need to test adding path prefixes.

Comment (alexandernanberg) [reply packages/gatsby-link/src/__tests__/index.js]: Sorry I'm sure that I'm following, don't we already have test for that in the bottom of the file?

**Comment by KyleAMathews** [comment]
Deploy preview for *gatsbyjs* failed.

Built with commit c843ac30b35e8baaf3a9c4491a0c3df73df0ea07

https://app.netlify.com/sites/gatsbyjs/deploys/5b694e9add28ef0e8b98aec3

**Comment by KyleAMathews** [comment]
Getting this error:

```
Summary of all failing tests
 FAIL  packages/gatsby-link/src/__tests__/index.js
  ● Test suite failed to run
    TypeError: Object.entries is not a function
      1 | import React from "react"
    > 2 | import { render, cleanup } from "react-testing-library"
      3 | import {
      4 |   createMemorySource,
      5 |   createHistory,
      
      at Object.<anonymous> (node_modules/dom-testing-library/dist/events.js:328:8)
      at Object.<anonymous> (node_modules/dom-testing-library/dist/index.js:68:15)
      at Object.<anonymous> (node_modules/react-testing-library/dist/index.js:10:26)
      at Object.<anonymous> (packages/gatsby-link/src/__tests__/index.js:2:1)=
```

It seems we're hitting this error https://github.com/kentcdodds/react-testing-library/issues/110

**Comment by KyleAMathews** [comment]
Seems easiest thing to do is import babel-polyfill?

**Comment by alexandernanberg** [comment]
Which node version are you using? `react-testing-library` requires at least a node version of 8 I believe

**Comment by KyleAMathews** [comment]
We test node 6, 8, and 10

**Comment by alexandernanberg** [comment]
Ah in that case adding the polyfill might be the best idea, although I suspect that more libraries will drop node 6 support pretty soon

**Comment by KyleAMathews** [comment]
Yeah, node 6 is in maintenance until early next year https://github.com/nodejs/Release

So we'll drop it sooner than later but no hurry.

**Comment by alexandernanberg** [comment]
Makes sense! Yay the tests are passing now (except the cypress one)
